### ISSUE TITLE
Handle routines returning `Result`

### ISSUE THREAD

**Issue by nsunderland1** [bheisler/criterion.rs#506]
I have some benchmarks structured something like this:

```rust
fn function_that_may_fail() -> Result<(), Error> {
    // ...
}

...

fn some_benchmark(c: &mut Criterion) {
    c.bench_function("function_that_may_fail", |b| {
        b.iter(|| function_that_may_fail());
    });
}
```

The problem with this is that if the `function_that_may_fail` fails, then I don't get any indication that it did. It's easy enough to `.expect` or `.unwrap` in every such case, but that's something that you can forget to do (especially if it's your first time writing Criterion benchmarks), and if you forget then it can lead to wildly incorrect timings when the function fails.

It'd be nice if Criterion made cases like this safer. Not sure what the best way to go about it without introducing a breaking change would be though.

**Comment by lemmih** [comment]
I'm not sure how you want criterion to handle errors? By aborting? Often you really do want to measure the performance of an action that fails so it's not clear-cut. As such, I think it's better for the user to deal with it.

**Comment by bheisler** [comment]
I agree, it would make more sense for the benchmark author (who has a better understanding of what their benchmark is supposed to do than I could have) to make decisions like this.

**Comment by nsunderland1** [comment]
I think my title and description may have distracted a bit from my real issue, that Criterion is effectively causing `must_use` values to be ignored, without warning. I agree that Criterion shouldn't be opinionated about `Result`, but I don't think it needs to be in order to avoid this issue. So really the root cause here is the signature of methods like [iter](https://docs.rs/criterion/0.3.5/criterion/struct.Bencher.html#method.iter):

```rust
pub fn iter<O, R>(&mut self, routine: R) where
    R: FnMut() -> O, 
```

In particular, `R: FnMut() -> O`. The return value of `O` is what causes the values of `must_use` types like `Result` to be ignored. In a hypothetical world where Criterion doesn't do this, the issue wouldn't exist:

```rust
pub fn iter<R>(&mut self, routine: R) where
    R: FnMut() -> (),
```

Of course, this is a huge breaking change and thus probably isn't a great idea, but it's the sort of thing that fixes the problem. With this signature, I can't do this:

```rust
b.iter(|| function_that_may_fail());
```

Because of the return type mismatch. So I'm forced to wrap it like this:

```rust
b.iter(|| {
    function_that_may_fail();
});
```

And then I get the warning about the `must_use`. It does come at the expense of one-liners, but it fixes the issue without making Criterion more opinionated. Once again though, super breaking change so I understand that this _particular_ approach probably isn't happening.

**Comment by lemmih** [comment]
I'm not sure I understand your point of view. I wouldn't say Criterion ignores results marked with `#[must_use]`. Rather, it calculates how long it takes to *generate* those results. Addition is also `#[must_use]` but benchmarking `x + y` is perfectly fine.

That you want to abort on `Err` results is a highly unusual workflow and I'm not sure I see the benefits. Feels like you're trying to do testing and benchmarking at the same time. Criterion shouldn't prevent you from doing this, of course, but neither should it go out of its way to support it, imo.

**Comment by tiye** [comment]
Got similar concern when I found one of my test actually overflowed with `u8`, the test runs and gives normal results. Is there a way for me to be noticed when it runs into error?

**Comment by RaulTrombin** [comment]
@tiye You can panic!()

**Comment by RaulTrombin** [comment]
I have an i2c sensor reading, and was not measuring if they are valid or get some error during the sampling.
It can shows a false upgrade performance on errors ( such not sensor attached or anything related, that will fail fast )

I just used unwrap_or_else with panic like follows:

```
    c.bench_function("new", |b| {
        b.iter(||
            sensor.read_register(Register::ST1).unwrap_or_else(|e| {
                panic!("Error during benchmark, reason: {:?}", e);
            })
        )
    });
```
